### ISSUE TITLE
fleet fails to move to backup server in the event of etcd failure

### ISSUE THREAD

**Issue by brianredbeard** [coreos/fleet#1157]
When fleet is passed a list of etcd servers it enumerates the servers specified in order on every single request. If the first server in the list has failed it will always first make an attempt to contact the first server which leads to needless network connectivity and an increase in load on the host.

Example:

```
$ cat /etc/fleet/fleet.conf
verbosity=4
etcd_servers="http://10.132.240.54:4001","http://10.132.240.40:4001","http://10.132.240.55:4001"
```

Logs:

```
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO engine.go:191: Lease already held by Machine(83dde1d9c93f43cf801e99848297254f) operating at acceptable version 1
Mar 20 01:19:17 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from PUT http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/machines/b67484e8b1b34d16a328a4dfaac545ec/object?prevExist=true: 200 OK
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:19 redbeard-2gb-3-3 fleetd[32102]: INFO engine.go:191: Lease already held by Machine(83dde1d9c93f43cf801e99848297254f) operating at acceptable version 1
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:21 redbeard-2gb-3-3 fleetd[32102]: INFO engine.go:191: Lease already held by Machine(83dde1d9c93f43cf801e99848297254f) operating at acceptable version 1
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&sorted=true
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&sorted=true: 404 Not Found
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&sorted=true
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&sorted=true: 404 Not Found
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:307: AgentReconciler attempting tasks []
Mar 20 01:19:22 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:63: AgentReconciler completed reconciliation in 35.647799ms
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:23 redbeard-2gb-3-3 fleetd[32102]: INFO engine.go:191: Lease already held by Machine(83dde1d9c93f43cf801e99848297254f) operating at acceptable version 1
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO reconcile.go:86: Reconciler tick
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/engine/version?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/lease/engine-leader?consistent=true&recursive=false&sorted=false: 200 OK
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO engine.go:191: Lease already held by Machine(83dde1d9c93f43cf801e99848297254f) operating at acceptable version 1
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: unexpected EOF
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.54:4001/: unexpected EOF
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: unexpected EOF
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.54:4001/: unexpected EOF
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.40:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.40:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.55:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: ERROR client.go:213: Unable to get result for {Watch /_coreos.com/fleet/job}, retrying in 100ms
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.55:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: ERROR client.go:213: Unable to get result for {Watch /_coreos.com/fleet/job}, retrying in 100ms
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.54:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.54:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.40:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.40:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.40:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.55:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: ERROR client.go:213: Unable to get result for {Watch /_coreos.com/fleet/job}, retrying in 200ms
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:373: etcd: recv response from GET http://10.132.240.55:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: 307 Temporary Redirect
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:291: Failed getting response from http://10.132.240.55:4001/: dial tcp 10.132.240.54:4001: connection refused
Mar 20 01:19:25 redbeard-2gb-3-3 fleetd[32102]: ERROR client.go:213: Unable to get result for {Watch /_coreos.com/fleet/job}, retrying in 200ms
Mar 20 01:19:26 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:366: etcd: sending HTTP request GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true
Mar 20 01:19:26 redbeard-2gb-3-3 fleetd[32102]: INFO client.go:369: etcd: recv error response from GET http://10.132.240.54:4001/v2/keys/_coreos.com/fleet/job?consistent=true&recursive=true&wait=true: dial tcp 10.132.240.54:4001: connection refused
```

**Comment by bcwaldon** [comment]
We should just replace fleet's etcd client with github.com/coreos/etcd/client

**Comment by bcwaldon** [comment]
This will be addressed by https://github.com/coreos/fleet/issues/1209

**Comment by jonboulle** [comment]
This landed via #1212
